### PR TITLE
move more operators to BatchRulesDecompositions

### PR DESCRIPTION
**Pull request by srossross**
Moving operators over to `BatchRulesDecompositions.cpp` to remove xfails. I noticed that composite-compliant does not mean inductor or vmap compliant, so I added more `isTensorSubclassLike` checks

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by lezcano — APPROVED** [review]
Nice clean up!

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp [hunk 1]
```diff
@@ -1474,7 +1474,7 @@ Tensor gather_backward(const Tensor& grad, const Tensor& self, int64_t dim, cons
   auto result = grad.new_zeros_symint(self.sym_sizes());
   // for composite compliance, use out-of-place variant of
   // `scatter_add` if index tensor is a Tensor Subclass.
-  if (isTensorSubclassLike(index)) {
+  if (isTensorSubclassLike(index) || isTensorSubclassLike(grad)) {
```
Comment (kshitij12345) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:1477]: minor nit: You can use `areAnyTensorSubclassLike`.

**Review by kshitij12345 — APPROVED** [review]
LGTM! Thank you @srossross

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp [hunk 2]
```diff
@@ -1474,7 +1474,7 @@ Tensor gather_backward(const Tensor& grad, const Tensor& self, int64_t dim, cons
   auto result = grad.new_zeros_symint(self.sym_sizes());
   // for composite compliance, use out-of-place variant of
   // `scatter_add` if index tensor is a Tensor Subclass.
```
Comment (kshitij12345) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:1476]: Also, it would make sense to update the comment

Comment (srossross) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: updated

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp (hunk 1 above)
Comment (srossross) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: changed

**Pull request closed without merging**